Auto detect python version and remove hard-coded version

Kushal-deb opened the pull request:
Auto detect python version and remove hard-coded versions

Fix Issue: https://tracker.ceph.com/issues/66903

Logs:

```
root@ceph-node-0 ~]# journalctl -n all -ft cloud-init
Nov 14 16:38:05 ceph-node-0 cloud-init[991]: Detected Python version 3.9 in container 5f50073063562b50c9f0c69e3627e21a2477bccfd503c4a692ef584adf2ddae8
Nov 14 16:38:05 ceph-node-0 cloud-init[991]: Retrieved python version matches the stored version; no update
```

```
[root@ceph-node-0 ~]# podman ps

CONTAINER ID  IMAGE                                                                                      COMMAND               CREATED         STATUS         PORTS       NAMES
5f5007306356  quay.io/kdeb/ceph:testing_auto-detect-1                                                    -n mon.ceph-node-...  24 minutes ago  Up 24 minutes              ceph-2c56d266-a2a6-11ef-bb23-525400e724b8-mon-ceph-node-0
04af53719729  quay.io/kdeb/ceph:testing_auto-detect-1                                                    -n mgr.ceph-node-...  24 minutes ago  Up 24 minutes              ceph-2c56d266-a2a6-11ef-bb23-525400e724b8-mgr-ceph-node-0-nqoiij
c8e4913a3fd0  quay.io/kdeb/ceph@sha256:fdae78b72cfa2f7996896b6dd504dfc93a53d4307af8563c1e8d75d654a01ba0  -n client.crash.c...  22 minutes ago  Up 22 minutes              ceph-2c56d266-a2a6-11ef-bb23-525400e724b8-crash-ceph-node-0
b4f4fdf34508  quay.io/kdeb/ceph@sha256:fdae78b72cfa2f7996896b6dd504dfc93a53d4307af8563c1e8d75d654a01ba0  -n osd.0 -f --set...  19 minutes ago  Up 19 minutes              ceph-2c56d266-a2a6-11ef-bb23-525400e724b8-osd-0
```

```
[root@ceph-node-0 ~]# podman exec -it ceph-2c56d266-a2a6-11ef-bb23-525400e724b8-mon-ceph-node-0 /bin/bash

[root@ceph-node-0 /]# cd /usr/lib/python3.9/site-packages/ceph

[root@ceph-node-0 ceph]# pwd
/usr/lib/python3.9/site-packages/ceph
```


==================================================================================================

```
❯ tox -e py39 -- cephadm
py39: commands[0]> pytest --doctest-modules cephadm
=========== test session starts =========
1243 passed, 1 skipped, 7 warnings in 337.29s (0:05:37) 
```

==================================================================================================
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

phlogistonjohn reviewed (changes requested):
please name your commit messages starting with the main topical area. in this case it would be `cephadm:`

I would really like to try avoiding growing cephadm.py whenever possible. Can you move the functionality into cephadmlib in an appropriate file?

rkachach commented on src/cephadm/cephadm.py:
```diff
@@ -2703,6 +2703,35 @@ def _rollback(ctx: CephadmContext) -> Any:
             raise
     return cast(FuncT, _rollback)
 
+def get_python_version(ctx: CephadmContext):
+    """
+    Retrieves the Python version from a specified daemon's container.
+    """
+    daemon_types = ['mon', 'mgr', 'osd']
+    setattr(ctx,'python_version',None)
+    for daemon in daemon_types:
+        container_info = get_container_info(ctx, daemon, by_name=False)
+        if container_info and container_info.container_id:
+            container_id = container_info.container_id
+            try:
+                exec_command = [
+                    "podman", "exec", container_id, "python3",
+                    "-c", "import sys; print(f'{sys.version_info.major}.{sys.version_info.minor}')"
+                ]
+                exec_result = subprocess.run(exec_command, capture_output=True, text=True)
+                python_version = exec_result.stdout.strip()
+        
+                # Store the detected Python version in ctx
+                if not ctx.python_version:
+                    ctx.python_version = python_version
+                else:
+                    if ctx.python_version == python_version:
+                        logger.info("Retrieved python version matches the stored version; no update")
+                    else:
+                        logger.warning("Python version mismatch")
+                logger.info(f"Detected Python version {python_version} in container {container_id}")
+            except Exception as e:
+                logger.error(f"Failed to get Python version for container {container_id}: {str(e)}")
```
As John already commented, there's no need to check all the daemons, you can use `mon` for example. Most of the time all of them will have the same version. I find the function a little bit complicated and hard to read, I'll suggest dividing it into two functions and have the main one call the other:

```
def get_python_version():
def retrieve_python_version_from_container(container_id):
```

Kushal-deb replied:
I have broken down the function into two parts now. Also it retrieves the python version from any one of the containers and returns instead of looping through all of them.

Kushal-deb commented on src/cephadm/cephadm.py:
```diff
@@ -2703,6 +2703,35 @@ def _rollback(ctx: CephadmContext) -> Any:
             raise
     return cast(FuncT, _rollback)
 
+def get_python_version(ctx: CephadmContext):
```
Addressed

Kushal-deb commented on src/cephadm/cephadm.py:
```diff
@@ -2858,7 +2887,7 @@ def cli(cmd, extra_mounts={}, timeout=DEFAULT_TIMEOUT, verbosity=CallVerbosity.V
 
     wait_for_mon(ctx, mon_id, mon_dir, admin_keyring.name, tmp_config.name)
-
+    
```
Removed

Kushal-deb commented on src/cephadm/cephadm.py:
```diff
@@ -2703,6 +2703,35 @@ def _rollback(ctx: CephadmContext) -> Any:
+            container_id = container_info.container_id
+            try:
+                exec_command = [
+                    "podman", "exec", container_id, "python3",
```
Addressed

Kushal-deb commented on src/cephadm/cephadm.py:
```diff
@@ -2703,6 +2703,35 @@ def _rollback(ctx: CephadmContext) -> Any:
+                if not ctx.python_version:
+                    ctx.python_version = python_version
+                else:
+                    if ctx.python_version == python_version:
```
Returning the value as soon as a valid one is retrieved.

Kushal-deb commented on src/cephadm/cephadmlib/daemons/ceph.py:
```diff
@@ -487,13 +487,13 @@ def get_ceph_mounts_for_type(
                 mounts[
                     ceph_folder + '/src/ceph-volume/ceph_volume'
-                ] = '/usr/lib/python3.9/site-packages/ceph_volume'
+                ] = f'/usr/lib/{ctx.python_version}/site-packages/ceph_volume'
```
Initialized a value in the constants.py file.
Please let me know if that's the correct way to do it?

Kushal-deb commented on src/cephadm/cephadm.py:
```diff
@@ -2986,6 +3015,8 @@ def mgr_has_latest_epoch():
                 '\thttps://docs.ceph.com/en/latest/mgr/telemetry/\n')
     logger.info('Bootstrap complete.')
 
+    get_python_version(ctx)
```
Moved the function to ceph.py where it is needed.

Kushal-deb commented:
Hi, @adk3798, Thanks for the detailed explanation.

I moved the ContainerInfo class to a new file under /cephadmlib. During the make check, I found that test_cephadm.py references this class as _cephadm.ContainerInfo, which caused a failure at that point.

when I ran tox -x, it highlighted an import error as I added an import statement in test_cephadm.py  to get the ContainerInfo class.

Below is the error: 

  ```
   from cephadmlib.daemons.ceph import ContainerInfo
E   ImportError: cannot import name 'ContainerInfo' from 'cephadmlib.daemons.ceph' (/home/kushaldeb/Github/ceph/src/cephadm/cephadmlib/daemons/ceph.py)
```
Are you allowed to import in test files?

adk3798 commented:
> Hi, @adk3798, Thanks for the detailed explanation.
> 
> I moved the ContainerInfo class to a new file under /cephadmlib. During the make check, I found that test_cephadm.py references this class as _cephadm.ContainerInfo, which caused a failure at that point.
> 
> when I ran tox -x, it highlighted an import error as I added an import statement in test_cephadm.py to get the ContainerInfo class.
> 
> Below is the error:
> 
> ```
>  from cephadmlib.daemons.ceph import ContainerInfo
> E   ImportError: cannot import name 'ContainerInfo' from 'cephadmlib.daemons.ceph' (/home/kushaldeb/Github/ceph/src/cephadm/cephadmlib/daemons/ceph.py)
> ```
> 
> Are we allowed to import in test files?

if we need to then yeah. For example in `src/cephadm/tests/test_host_facts.py` it imports an `Enclosure` object from `src/cephadm/cephadmlib/host_facts.py` as it needs to use that object for the tests.

Kushal-deb commented:
jenkins test make check

phlogistonjohn reviewed (changes requested):
Please do not refactor and make behavior changes in the same commit (same PR is ok).
It makes it nearly impossible for future readers to disentangle what changes involved moving things around versus what is net new. If I or someone else lead you to think that putting everything in a single commit was desired, sorry. (If it's easier for you, then 2 separate PRs is fine... it's just not mandatory)

I would also discourage putting things under cephadmlib/daemons that are not daemon implementations. I would also discourage naming the new file "utils" unless it's truly a bunch of somewhat unrelated stuff. Naming is hard though and I won't insist on it. If the theme of the new file is listing or querying daemons try names like "daemon_info" or "daemon_list" or "daemon_state" or so on... feel free to break out the thesaurus :-)

ceph-jenkins commented:
Can one of the admins verify this patch?

Kushal-deb committed:
cephadm: Auto detect python version

Signed-off-by: Kushal Deb <Kushal.Deb@ibm.com>

Kushal-deb commented:
> Please do not refactor and make behavior changes in the same commit (same PR is ok). It makes it nearly impossible for future readers to disentangle what changes involved moving things around versus what is net new. If I or someone else lead you to think that putting everything in a single commit was desired, sorry. (If it's easier for you, then 2 separate PRs is fine... it's just not mandatory)
> 
> I would also discourage putting things under cephadmlib/daemons that are not daemon implementations. I would also discourage naming the new file "utils" unless it's truly a bunch of somewhat unrelated stuff. Naming is hard though and I won't insist on it. If the theme of the new file is listing or querying daemons try names like "daemon_info" or "daemon_list" or "daemon_state" or so on... feel free to break out the thesaurus :-)

Thank you for the detailed feedback; it’s much appreciated!

Previously, I attempted to refactor some parts of the code while implementing the behavior changes in the same commit. However, this caused numerous issues with the mocks during testing, where fixing one issue often led to failures in others. To avoid these problems and ensure cleaner implementation, I’ve reverted the refactoring for now and kept the focus of this PR solely on the behavior change. I plan to take up the refactoring work gradually as part of a separate PR to avoid introducing further complexity or regressions.

For this PR, I’ve opted to use a lazy import approach to detect the Python version for the time being. This allows us to move forward without touching too many existing structures, keeping the changes minimal and isolated.

phlogistonjohn commented on src/cephadm/cephadmlib/daemons/ceph.py:
```diff
@@ -407,6 +410,55 @@ def create_daemon_dirs(self, data_dir: str, uid: int, gid: int) -> None:
         populate_files(ssl_dir, cert_files, uid, gid)
 
 
+def retrieve_version(container_id: str, container_command: str) -> str:
```
Since this function is only used within this file naming it `_retrieve_version` would be cleaner as it says 'not part of the module's public api'.

phlogistonjohn commented on src/cephadm/cephadmlib/daemons/ceph.py:
```diff
@@ -407,6 +410,55 @@ def create_daemon_dirs(self, data_dir: str, uid: int, gid: int) -> None:
+            "import sys; print(f'{sys.version_info.major}.{sys.version_info.minor}')",
+        ]
+
+        exec_result = subprocess.run(
+            exec_command, capture_output=True, text=True, check=True
+        )
```
For consistency with other parts of cephadm I think you should use one of the functions in `call_wrappers` library.

phlogistonjohn commented on src/cephadm/cephadmlib/daemons/ceph.py:
```diff
@@ -407,6 +410,55 @@ def create_daemon_dirs(self, data_dir: str, uid: int, gid: int) -> None:
+    return ''
+
+
+def get_python_version(ctx: CephadmContext) -> str:
```
Since this function is only used within this file naming it `_get_python_version` would be cleaner as it says 'not part of the module's public api'.

phlogistonjohn commented on src/cephadm/cephadmlib/daemons/ceph.py:
```diff
@@ -407,6 +410,55 @@ def create_daemon_dirs(self, data_dir: str, uid: int, gid: int) -> None:
+
+def get_python_version(ctx: CephadmContext) -> str:
+    try:
+        from cephadm import get_container_info
```
This is a little odd. I know you're trying to work around a dependency issue but I'm not sure if this is the best way to go about it.  However, I'm not going to outright block this PR over if if we can make this a temporary thing... let's think on it some more.

phlogistonjohn replied:
FWIW, this patch as well as worked I started for https://tracker.ceph.com/issues/69278 is what motivated this refactor. an initial PR should be up very soon.

adk3798 replied:
@Kushal-deb FYI the patches we were talking about in this thread have been merged, so you could take another look at this when you have bandwidth

phlogistonjohn reviewed (commented):
Looks much improved. My only major issue now is the import trick.

phlogistonjohn commented:
For the next time you update the PR. It is good to follow ceph's recommended practice for commit message titles: https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst#describe-your-changes
